sparta/sparta/memory/DMIBlockingMemoryIF.hpp

```diff
@@ -1,84 +1,100 @@
 
 #pragma once
 
-#include "sparta/memory/DMI.hpp"
 #include "sparta/memory/BlockingMemoryIF.hpp"
 
 namespace sparta {
-namespace memory {
+    namespace memory {
 
-/**
- * \class DMIBlockingMemoryIF
- * \brief Wrapps a DMI in a BlockingMemoryIF
- */
-class DMIBlockingMemoryIF final : public BlockingMemoryIF
-{
-public:
-    DMIBlockingMemoryIF(const DMI &dmi)
-    : BlockingMemoryIF("DMI",
-                       dmi.getSize(),
-                       DebugMemoryIF::AccessWindow(
-                           dmi.getAddr(), dmi.getAddr() + dmi.getSize()))
-    , dmi_(dmi)
-    {
-    }
+        /**
+         * \class DMIBlockingMemoryIF
+         * \brief Class that provides DMI access to backend memory
+         */
+        class DMIBlockingMemoryIF final : public BlockingMemoryIF
+        {
+        public:
+            /**
+             * \brief Wraps a raw pointer and provides BlockingMemoryIF to it
+             * \param raw_pointer The raw data pointer with memory to access
+             * \param start_addr  The expected "start address" of this raw pointer
+             * \param size        The expected size of the data
+             *
+             * Typically created by BlockingMemoryObjectIFNode when
+             * DMI request is made, this class will wrap a raw data
+             * pointer and provide the functionality of the
+             * BlockingMemoryIF.  This includes window access checking
+             * and bounds checking for the raw pointer.
+             */
+            DMIBlockingMemoryIF(void * raw_pointer,
+                                addr_t start_addr,
+                                addr_t size) :
+                BlockingMemoryIF("DMI", size, {start_addr, size}),
+                start_addr_(start_addr),
+                raw_pointer_(raw_pointer)
+            { }
 
-    addr_t getAddr() const
-    {
-        return dmi_.getAddr();
-    }
+            //! \return The internal raw pointer.
+            //!
+            //! This is dangerous to use directly as memory bounds
+            //! checking can easily be bypassed
+            void * getRawDataPtr() { return raw_pointer_; }
```

Comment:
I think the assumption made for raw pointer is that the data layout of raw data has to be continuous, though I can't come up with a case which isn't.

Reply:
It's guaranteed to be continuous _only within the BLOCK_SIZE_.  I can update the comments.  If you try to access memory using this pointer beyond the block size, you're gonna find trouble.

Reply:
Big trouble!